src/devenv/cli.py

```diff
@@ -0,0 +1,689 @@
+    ctx.ensure_object(dict)
+    ctx.obj['system'] = system
+    ctx.obj['verbose'] = verbose
+    ctx.obj['command_flags'] = []
+    ctx.obj['nix_flags'] = list(nix_flags)
+    ctx.obj['nix_flags'] += ['--system', system]
+    if offline:
+        ctx.obj['nix_flags'] += ['--offline']
+    if debugger:
+        # ignore-try is needed to avoid catching unrelated errors
+        ctx.obj['command_flags'] += ['--debugger', '--ignore-try']
+        # to avoid confusing errors
+        disable_eval_cache = True
+    if disable_eval_cache:
+        ctx.obj['nix_flags'] += ['--option', 'eval-cache', 'false']
+
+    if 'XDG_DATA_HOME' not in os.environ:
+        ctx.obj['gc_root'] = os.path.join(os.environ['HOME'], '.devenv', 'gc')
+    else:
+        ctx.obj['gc_root'] = os.path.join(os.environ['XDG_DATA_HOME'], 'devenv', 'gc')
+    
+    Path(ctx.obj['gc_root']).mkdir(parents=True, exist_ok=True)
+    ctx.obj['gc_project'] = os.path.join(ctx.obj['gc_root'], str(int(time.time() * 1000)))
+
+@cli.group()
+def processes():
+    pass
+
+
+DEVENV_DIR = Path(os.getcwd()) / '.devenv'
+os.environ['DEVENV_DIR'] = str(DEVENV_DIR)
+DEVENV_GC = DEVENV_DIR / 'gc'
+os.environ['DEVENV_GC'] = str(DEVENV_GC)
+
+PROCESSES_PID = DEVENV_DIR / 'processes.pid'
+PROCESSES_LOG = DEVENV_DIR / 'processes.log'
+
+
+def add_gc(name, store_path):
+    """Register a GC root"""
+    ctx = click.get_current_context()
+    run_command(f'nix-store --add-root "{os.environ["DEVENV_GC"]}/{name}" -r {store_path} >/dev/null')
+    os.symlink(store_path, f'{ctx.obj["gc_project"]}-{name}', True)
+
+
+@cli.command(hidden=True)
+@click.pass_context
+def assemble(ctx):
+    if not os.path.exists('devenv.nix'):
+        log('File devenv.nix does not exist. To get started, run:', level="error")
+        log('  $ devenv init', level="error")
+        exit(1)
+
+    DEVENV_GC.mkdir(parents=True, exist_ok=True)
+
+    if os.path.exists('devenv.yaml'):
+        validate_and_parse_yaml(DEVENV_DIR)
+    else:
+        for file in ['devenv.json', 'flake.json', 'imports.txt']:
+            file_path = DEVENV_DIR / file
+            if file_path.exists():
+                os.remove(file_path)
+
+    with open(FLAKE_FILE_TEMPL) as f:
+        flake = f.read()
+        system = ctx.obj['system']
+
+        with open(FLAKE_FILE, 'w') as f:
+            devenv_vars = (f"""
+  version = "{get_version()}";
+  system = "{system}";
+  devenv_root = "{os.getcwd()}";
+            """)
+            # replace __DEVENV_VARS__ in flake using regex
+            flake = re.sub(r'__DEVENV_VARS__', devenv_vars, flake)
+            f.write(flake)
+
+
+@cli.command(
+    help="Deletes previous devenv generations. See http://devenv.sh/garbage-collection",
+    short_help="Deletes previous devenv generations. See http://devenv.sh/garbage-collection",
+)
+@click.pass_context
+def gc(ctx):
+    GC_ROOTS = ctx.obj['gc_root']
+    start = time.time()
```

Comment:
Not that it would matter much here, but there is also time.perf_counter() for time measurements.